### PR TITLE
Database Tables

### PR DESCRIPTION
**Pull request by WyrdWyn4**
### Ticket: TT-50 & TT-57

# Changes
## 1. `Dockerfile`
 - Removed COPY `event-service` in the docker file, as event-service folders have been removed, but the docker file wasn't updated. This caused the docker file to not be built.
## 2. `Main.go`
 - Instead of calling individual functions for Deleting and Creating tables, I created one function to delete all tables, and create all tables, and this is now called in `main.go`.
## 3.  `Database.go`
Added the following functions and implementations:
 - ### CreateAllTables()
    - This function calls the create functions for each table.
 - ### DeleteAllTables()
     - This function drops all tables.
 - ### IntsToStrings() and StringsToInts()
     - These functions convert to and fro between Ints and Strings
 - ### UpdateAttribute()
     - Instead of creating individual functions for each and every attribute update, I created a general helper function to allow us to do the same task.
     - Through the use of interfaces, different types can be passed into this function.
 - ### AddAttributeArray()
     - In most tables, we are storing IDs as arrays to represent one-to-many or many-to-many relationships. This is a helper function that allows us to Add values to the existing arrays of data.
     - Again, through the use of interfaces, different types can be passed into this function. However, there are issues with reading data other than Strings, which is why I haven't held any other array-type implementations.
 - ### RemoveAttributeArray()
     - Just like the above, this function removes IDs from arrays, rather than add them.
## 3.  `User-db.go`
 - ### Column Changes
     - Added Profile columns to the User table _(Name, Country, Languages, Tags)_
     - Added Front-linking array columns to the User table _(Boards and Posts)_
 - ### Function Changes
     - Updated `CreateUserTable()` function
     - `AddUser()` now takes a struct as an argument rather than column values
     - ~`UpdateUser()`~ is now outdated and has been replaced with individual update functions for each column, which call upon `UpdateAttribute()`, `AddAttributeArray()`, and `RemoveAttributeArray()` respectively.
     - User has a One-to-many relationship with Boards. If a Board gets deleted, it will be removed from the User's list of boards, and this logic has been handled in the code.
     - User has a One to Many relationship with Posts. If a Post gets deleted, it will be removed from the User's list of boards, and this logic has been handled in the code.
     - If the User is deleted, all Posts and Boards will also be deleted.
## 4.  `Board-db.go`
 - ### Column Changes
     - Added columns _(BoardId, Name, CreationDate, Description, Username, Posts, Tags)_
 - ### Functionality Changes
     - Created Functionality for CRUD Operations.
     - Board has a Many to Many relationship with Posts _(One board can contain many Posts, and the same Post can be on many boards)_
## 5.  `Post-db.go`
 - ### Column Changes
     - Added columns _(PostId, ItineraryId, Title, ImageLink, Description, CreationDate, Username, Tags )_
 - ### Functionality Changes
     - Created Functionality for CRUD Operations.
     - Post has a Many to Many relationship with Board _(One board can contain many Posts, and the same Post can be on many boards)_. If a Post is deleted, it will be removed from all Boards, and this logic has been handled in the code.
     - Post has a One to One relationship with Itinerary. In fact, it would be considered a subset. Every Post has an itinerary, but not every Itinerary has a post. If an Itinerary gets deleted, the post also gets deleted, and this logic has been handled in the code.
## 6.  `Itinerary-db.go`
 - ### Column Changes
     - Added columns _(ItineraryId, Name, Country, Languages, Tags, Events, PostId, Username, CreationDate, LastUpdate )_
 - ### Functionality Changes
     - Created Functionality for CRUD Operations.
     - Itinerary has a One to One relationship with Post. In fact, it would be considered a subset. Every Post has an itinerary, but not every Itinerary has a post. If an Itinerary gets deleted, the post also gets deleted, and this logic has been handled in the code.
     - Event has a Many to Many relationship with Itineraries. An Itinerary may include multiple events, whereas an event could be a part of multiple itineraries.
## 7.  `Event-db.go`
 - ### Column Changes
     - Added columns _(EventId, Name, Price, Location, Description, StartDate, EndDate, ItineraryIds, PhotoLinks)_
 - ### Functionality Changes
     - Created Functionality for CRUD Operations.
     - Event has a Many to Many relationship with Itineraries. An Itinerary may include multiple events, whereas an event could be a part of multiple itineraries.
## 8.  `Authentication Service`
 - I refactored @AbdulTur's code to work with the updated database tables.

# Notes
 - Since our application use-case is small, I have added linking columns for relationships to avoid complexities with Join tables. Basically, each table will have a tracker linking column for its related table.
   - For example, a Board might contain several Posts, so it will contain a list of those Posts. This is called Frontlinking. 
   - Next, a Post might be contained in several Boards, so it will contain a list of those Boards. This is called Backlinking.
   - Whenever we have any updates to either a post or board, we would need to use these linking columns to apply those changes in the related tables as well.
 - The issue with linking columns is that we wish to store an array of integers (primary key Ids). However, while we can store an array of Integers in PostgreSQL, the scanner has issues reading those integers out. For this reason, we will store the numbers as strings, and when we pull them out, we will convert them back into integers. To help with this, we have the IntsToStrings() and StringsToInts() functions.
 - We need to solve the concurrency issue with the backend starting before the database, as our docker files are getting bigger and bigger, this issue is getting worse. I have to rerun the `docker-compose up --build` command twice almost 90% of the time.
 - While writing out the logic for these, I was referring to our _6400 Assignment 1 - UML Diagram_. However, some things just didn't make sense. For example:
   - The user has direct relationships with posts and boards but not with itineraries. Isn't the source of the itineraries the user himself? How is the user going to create these itineraries?
     - What is the source of Itineraries and Events? Are events just going to be added by the program itself?
     - If the user is the source of all of these things, then we not only need a new UML Diagram, but we also need to refactor our tables for added linking columns and our functionalities for linked column updates.
   - How much effect does the source table have on the drain table?
     - If a user is deleted, do we delete all of their boards and posts?
     - If an itinerary is deleted, do we delete the post as well?
   - What happens to itineraries that are not posted? What's the point of having them in the app? You can not save them to boards (According to our UML Diagram, only a post can be saved to a board)
 - I added Recursive Logic for the Add functions.
   - Basically, if you add a Post to the list of posts in a Board, it should also add the Board to a list of boards in the post.
   - However, if you call the AddPost function in the AddBoard function, and call the AddBoard function in the AddPost function, it leads to infinite recursion.
   - To solve this issue, I added an additional argument, which is recursion = false, to stop that.
   - Whenever we're using any AddPost, AddBoard, etc. functions, we need to set recursion = true so that both tables get updated.
 - Philosophical question, but what the hell is a post?

### CODE REVIEW

**Comment by AbdulTur** [comment]
@WyrdWyn4  Have you tested the signup and signin features and if they are still working fine, logically they should be but just in case?

**Comment by AbdulTur** [comment]
As far as i understood its through post we create iteneraries but i could be wrong, I also need a refressher of the link between a post itenary and board @jordyob03

**Comment by AbdulTur** [comment]
made the fix in 8f89dc191b08d442c5b552f314d7655df039b9f3 
if u have any questions on why the changes were made feel free to message me, but generally speaking the gin structs are there only for requests and if the stuff that sent from frontend mismatches  the structure of that struct errors are gonna be thrown

**Review by AbdulTur — APPROVED** [review]
LGTM

**Comment by Leo-Gilbert** [comment]
About point 6. I think we need an addition Column for city which contains one or multiple cities for searching purposes.

Also, I am confused about the idea of a linking column - how is this better than joining tables (I'm not doubting you, just genuinely do not understand

**Comment by AbdulTur** [comment]
> Firstly, you're right about multiple cities, but I'm thinking, should we just make it so that each itinerary has one city, to reduce complexity? Otherwise, we'd have to store an array of strings, which is doable, but idk, it'll probably be useless complication.

I think for now we said its 1 city just for the MVP but later we will see, if we do multiple it will indeed be string

**Comment by WyrdWyn4** [comment]
> > Firstly, you're right about multiple cities, but I'm thinking, should we just make it so that each itinerary has one city, to reduce complexity? Otherwise, we'd have to store an array of strings, which is doable, but idk, it'll probably be useless complication.
> 
> I think for now we said its 1 city just for the MVP but later we will see, if we do multiple it will indeed be string

Yeah, tbh, it will be a very quick update. I can probably get it done in 10 minutes, if you guys want. However, idk if @Leo-Gilbert started work already with just one city in mind.

**Comment by AbdulTur** [comment]
> > > Firstly, you're right about multiple cities, but I'm thinking, should we just make it so that each itinerary has one city, to reduce complexity? Otherwise, we'd have to store an array of strings, which is doable, but idk, it'll probably be useless complication.
> > 
> > 
> > I think for now we said its 1 city just for the MVP but later we will see, if we do multiple it will indeed be string
> 
> Yeah, tbh, it will be a very quick update. I can probably get it done in 10 minutes, if you guys want. However, idk if @Leo-Gilbert started work already with just one city in mind.



I think 1 city is fine for now for the MVP, good job on the work

**Review by jordyob03 — APPROVED** [review]
LGTM
